lib/ClangImporter/ImportDecl.cpp

```diff
@@ -3804,26 +3835,44 @@ namespace {
       if (!importedType)
         return nullptr;
 
-      auto resultTy = importedType.getType();
       auto loc = Impl.importSourceLoc(decl->getLocation());
 
       // FIXME: Poor location info.
       auto nameLoc = Impl.importSourceLoc(decl->getLocation());
-      result = createFuncOrAccessor(Impl.SwiftContext, loc, accessorInfo, name,
-                                    nameLoc, bodyParams, resultTy,
-                                    /*throws*/ false, dc, decl);
-
-      if (!dc->isModuleScopeContext()) {
-        if (selfIsInOut)
-          result->setSelfAccessKind(SelfAccessKind::Mutating);
-        else
-          result->setSelfAccessKind(SelfAccessKind::NonMutating);
-        if (selfIdx) {
-          result->setSelfIndex(selfIdx.getValue());
-        } else {
-          result->setStatic();
-          result->setImportAsStaticMember();
+
+      AbstractFunctionDecl *result = nullptr;
+      if (auto *ctordecl = dyn_cast<clang::CXXConstructorDecl>(decl)) {
+        // TODO: Is failable, throws etc. correct?
+        DeclName ctorName(Impl.SwiftContext, DeclBaseName::createConstructor(),
+                          bodyParams);
+        result = Impl.createDeclWithClangNode<ConstructorDecl>(
+            decl, AccessLevel::Public, ctorName, loc, /*failable=*/false,
+            /*FailabilityLoc=*/SourceLoc(), /*Throws=*/false,
+            /*ThrowsLoc=*/SourceLoc(), bodyParams, /*GenericParams=*/nullptr,
+            dc);
```

Comment:
Does this mean the Clang Importer will incorrectly translate constructors that throw on failure to non-throwing constructors? That doesn't sound right... I can understand if you want to punt it to a later time though, but please file a JIRA (and link that here) if you are doing that.

Reply:
I think that adding a test case for the problem would also be a good idea.

Reply:
> Does this mean the Clang Importer will incorrectly translate constructors that throw on failure to non-throwing constructors? That doesn't sound right...

Agree, this definitely shouldn't happen. I plan to resolve the `TODO` one way or the other before this PR is ready for review (I've put the PR up here for the time being to get feedback on the general approach).

I think what I'll probably do initially is to only import C++ constructors that are marked `noexcept`. Importing throwing C++ constructors (and mapping them to throwing Swift initializers) would then come as a second step. In this case, I assume filing a Jira ticket would not be necessary in this case (as we are then talking only about missing functionality, not erroneous functionality)?

And yes, I will definitely add test cases for this.

Reply:
I've added a commit (with tests) that restricts imports only to `noexcept` constructors for the time being.

Reply:
I've thought about exception handling some more and have written up a separate post for it:

https://forums.swift.org/t/handling-c-exceptions/34823/2

tl;dr: I don't think potentially-throwing C++ constructors (which really just means C++ constructors not marked `noexcept`) should be mapped to `throws` initializers because it results in non-idiomatic C++ code. Many C++ constructors don't in fact throw, but aren't explicitly marked `noexcept`; the same is true for functions, and indeed, we currently import potentially-throwing C++ functions as non-throwing Swift functions. Instead, the default behavior should be to import C++ functions as non-throwing; in case the C++ function does throw, we terminate the program.

Reply:
>  In this case, I assume filing a Jira ticket would not be necessary (as we are then talking only about missing functionality, not erroneous functionality)?

Yes, that's correct. Thanks for starting the broader discussion.

Reply:
As per the discussion on the forum, we want to import all C++ constructors, whether they are marked `noexcept` or not, as non-throwing initializers. I've therefore reverted the change that only imported `noexcept` constructors.